### ISSUE TITLE
MQTT plugin retained messages stream memory leak

### ISSUE THREAD

**Issue by mtaobiz** [nats-io/nats-server#3745]
## Defect

$MQTT_sess stream contains subjects each of them mapped to mqtt topic. 
$MQTT_rmsgs stream stores retained messages all on one subject with unlimited depth.
When new retained message received. its new data stored in rmsgs and old data removed from stream. 
Number of messages in $MQTT_rmsgs stream must be not more than 2 x number of topics with retained messages (2 messages per topic: old and new)

```
#nats str info \$MQTT_sess -j
{
  "config": {
    "name": "$MQTT_sess",
    "subjects": [
      "$MQTT.sess.\u003e"
    ],
    "retention": "limits",
    "max_consumers": -1,
    "max_msgs_per_subject": 1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "duplicate_window": 120000000000,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  },
  "created": "2022-12-23T06:10:07.483116925Z",
  "state": {
    "messages": 2129,
    "bytes": 418539,
    "first_seq": 45,
    "first_ts": "2022-12-23T06:10:10.134652266Z",
    "last_seq": 492084,
    "last_ts": "2022-12-27T12:23:22.98920217Z",
    "num_deleted": 489911,
    "num_subjects": 2129,
    "consumer_count": 0
  }
}
```
sess state.num_subjects = 2129
expected number of messages in rmsgs stream (state,messages) not more than: 2129 x 2 = 4258
```
# nats str info \$MQTT_rmsgs -j
{
  "config": {
    "name": "$MQTT_rmsgs",
    "subjects": [
      "$MQTT.rmsgs"
    ],
    "retention": "limits",
    "max_consumers": -1,
    "max_msgs_per_subject": -1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "duplicate_window": 120000000000,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  },
  "created": "2022-12-23T06:10:07.484662983Z",
  "state": {
    "messages": 8304,
    "bytes": 2418002,
    "first_seq": 480664084,
    "first_ts": "2022-12-27T12:04:45.979404197Z",
    "last_seq": 482360254,
    "last_ts": "2022-12-27T12:24:03.321903836Z",
    "num_deleted": 1687867,
    "num_subjects": 1,
    "consumer_count": 1
  }
}
```
but i see state.messages = 8304

few minutes later:
```
# nats str info \$MQTT_rmsgs -j
{
  "config": {
    "name": "$MQTT_rmsgs",
    "subjects": [
      "$MQTT.rmsgs"
    ],
    "retention": "limits",
    "max_consumers": -1,
    "max_msgs_per_subject": -1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "duplicate_window": 120000000000,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  },
  "created": "2022-12-23T06:10:07.484662983Z",
  "state": {
    "messages": 8466,
    "bytes": 2465974,
    "first_seq": 480664084,
    "first_ts": "2022-12-27T12:04:45.979404197Z",
    "last_seq": 482404917,
    "last_ts": "2022-12-27T12:24:45.451499771Z",
    "num_deleted": 1732368,
    "num_subjects": 1,
    "consumer_count": 1
  }
}
```
state.messages = 8466
next:
```
# nats str info \$MQTT_sess -j | jq .state.num_subjects
2137
# nats str info \$MQTT_rmsgs -j | jq .state.messages
12599
# nats str info \$MQTT_sess -j | jq .state.num_subjects
2137
# nats str info \$MQTT_rmsgs -j | jq .state.messages
13919
```

And it continues grow and finally i receive error: "unable to store retained message for account "$G", subject "...": insufficient resources (10023)"

#### Versions of `nats-server` and affected client libraries used:
nats version: 2.8.4

#### OS/Container environment:
SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux

#### Steps or code to reproduce the issue:
Continuous pulishing retained messages to mqtt topics

#### Expected result:
Jetstream MQTT retained stream storage not grow.

#### Actual result:
Jetstream MQTT retained stream storage grow.

**Comment by derekcollison** [comment]
Could you verify if you see the same behavior on the latest server? 2.9.10?

We will be updating the retained messages architecture for release 2.10 to improve.

**Comment by derekcollison** [comment]
I believe I understand what is going on, and yes this should be improved for sure. Will be for 2.10.

**Comment by mtaobiz** [comment]
this is in 2.9.10:

```
# nats str info \$MQTT_sess -j
{
  "config": {
    "name": "$MQTT_sess",
    "subjects": [
      "$MQTT.sess.\u003e"
    ],
    "retention": "limits",
    "max_consumers": -1,
    "max_msgs_per_subject": 1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "duplicate_window": 120000000000,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  },
  "created": "2022-11-30T14:11:13.972044267Z",
  "state": {
    "messages": 1380,
    "bytes": 182064,
    "first_seq": 1,
    "first_ts": "2022-11-30T14:11:13.977722361Z",
    "last_seq": 659379,
    "last_ts": "2023-01-02T06:11:19.618070722Z",
    "num_deleted": 657999,
    "num_subjects": 1380,
    "consumer_count": 0
  }
}
# nats str info \$MQTT_rmsgs -j
{
  "config": {
    "name": "$MQTT_rmsgs",
    "subjects": [
      "$MQTT.rmsgs"
    ],
    "retention": "limits",
    "max_consumers": -1,
    "max_msgs_per_subject": -1,
    "max_msgs": -1,
    "max_bytes": -1,
    "max_age": 0,
    "max_msg_size": -1,
    "storage": "file",
    "discard": "old",
    "num_replicas": 1,
    "duplicate_window": 120000000000,
    "sealed": false,
    "deny_delete": false,
    "deny_purge": false,
    "allow_rollup_hdrs": false,
    "allow_direct": false,
    "mirror_direct": false
  },
  "created": "2022-11-30T14:11:13.975680681Z",
  "state": {
    "messages": 1759703,
    "bytes": 517546717,
    "first_seq": 8172084144,
    "first_ts": "2022-12-30T06:23:45.553199969Z",
    "last_seq": 8715227923,
    "last_ts": "2023-01-02T06:11:30.138094599Z",
    "num_deleted": 541384077,
    "num_subjects": 1,
    "consumer_count": 1
  }
}
```

**Comment by derekcollison** [comment]
Thanks, will see if we can improve a bit for a possible 2.9.11 release, but will re-architect for 2.10, which should come in Jan sometime.

**Comment by derekcollison** [comment]
I just did a quick test and this seems to be working correctly. Every message published with a retain flag set will be retained. If another message with the same subject is sent also with a retained flag the original retained messages will be replaced by this one.

My simple test was to send 100 messages on subject `foo/#`, e.g `foo/1, foo/2`.

I then repeated that loop again and checked that the `$MQTT_rmsgs` was only holding 100 messages, with sequences starting at 100 through 200.
